### ISSUE TITLE
Potential bug

### ISSUE THREAD

**Issue by pnlfang** [ebimodeling/biocro#4]
Line 227 " Ci = Ca - (Assim * 1e-6 * 1.6 * AP) / (Gs * 0.001)" in c3photo.c seems to be a problem.  This gives Ci in Pa, while the rest of the code use ppm (umol mol-1).  I could be wrong, but please check the unit consistency.

**Comment by justinmcgrath** [comment]
Something seems to be wrong with the units. c3photoC appears to expect Ca to have units of Pa, but c3photoParms gives units of micromole / mol for Catm, which is eventually passed to c3photoC.

In comparison, c4photoC accepts Catm as micromole / mol, and in line 231 converts the units to Pa, which is then used in line 294 in an equation similar to that used in c3photoC.

@djaiswal, @danwang2008  or @femiguez can you confirm that the usage in c3photo is a mistake?

**Comment by justinmcgrath** [comment]
The 10 would appear to be an approximation to convert from Pa to micromole / mol. I'll change that line.

**Comment by pnlfang** [comment]
Another issue on line 227 in c3photo.c: is there a reason to use Ca rather than Cs to calculate Ci?  This could make a big difference in Ci calculated.

**Comment by justinmcgrath** [comment]
> Another issue on line 227 in c3photo.c: is there a reason to use Ca rather than Cs to calculate Ci? This could make a big difference in Ci calculated.

Agreed. [CO2] at the leaf surface should be used.

**Comment by justinmcgrath** [comment]
@pnlfang, I added a new branch, new-framework, to the repository. Could you try using this version to see if it meets your needs? I'd like make changes to this new framework rather than trying to keep the two frameworks working in parallel.

**Comment by justinmcgrath** [comment]
The README.md file gives the basic usage. A lot of the old functions have been removed though. Some will be rewritten and replaced, but many of them will be implemented differently.

**Comment by pnlfang** [comment]
@justinmcgrath Thanks.  I'll give it a try.  I noticed most of the parameters such as Kc, Ko etc. were taken from BERNACCHI et al. (1991), which were derived for tobacco.  Are they suitable for c3 grass?

**Comment by justinmcgrath** [comment]
> @justinmcgrath Thanks. I'll give it a try. I noticed most of the parameters such as Kc, Ko etc. were taken from BERNACCHI et al. (1991), which were derived for tobacco. Are they suitable for c3 grass?

Kinetics of isoforms from c3 plants are likely similar based on selective pressure on Rubisco, so the values from tobacco are probably acceptable for another c3 plant. Doug Orr surveyed a lot of isoforms though, and I'm pretty certain there are c3 grasses in the data set. You can get the values at 25 degrees C from that (10.1104/pp.16.00750). The data set doesn't have quite what you'd need for a proper temperature response, so you'd still have to use the values from tobacco for the Arrhenius parameters.

Moving those parameters out of the hard code and into the parameter set is on the to-do list.

**Comment by pnlfang** [comment]
@justinmcgrath Thank you very much!

**Comment by pnlfang** [comment]
@justinmcgrath I tested the code under the new framework, I noticed you made some changes.  Calculated Ci in c3photo.c now has unit of Pa, but Gstar has units of umol.  The unit conversion was not properly coded.
An unrelated question - if I have measured An, gs, and Ci, how confident to use the following equation to calculate the leaf boundary layer conductance: Ci = Ca – An(1.4/gb + 1.6/gs)?

**Comment by justinmcgrath** [comment]
> @justinmcgrath I tested the code under the new framework, I noticed you made some changes. Calculated Ci in c3photo.c now has unit of Pa, but Gstar has units of umol. The unit conversion was not properly coded.

It seems that code from c4photo.c was copied when making c3photo.c, and the differences in units between the models were not considered. I'll go through each function and carefully check the units.

> An unrelated question - if I have measured An, gs, and Ci, how confident to use the following equation to calculate the leaf boundary layer conductance: Ci = Ca – An(1.4/gb + 1.6/gs)?

Long and Hallgren equation 6.17 use roughly that equation (1.37 instead of 1.4). That's appropriate if gb is conductance of both sides of the leaf. If it's conductance of only one side, it needs to be adjusted.

**Comment by pnlfang** [comment]
Could you be more specific as to how to adjust if it's conductance of only one side?

**Comment by justinmcgrath** [comment]
> Could you be more specific as to how to adjust if it's conductance of only one side?

If the stomatal conductance is the same on both sides, then you multiply the one-sided boundary conductance by two. If the stomatal conductance is different on the two sides, which can happen because many leaves have more stomata on the abaxial surface, then the approach LiCOR uses is to adjust by
```
kf = (K^2 + 1) / (K + 1)^2
gt = 1 / (1.6 / gs + 1.37 * kf / gb)
```
where K is the ratio of stomatal conductance on one side to the conductance on the other side. When K = 1 then kf = 1/2.

Since K is often not known, when using a one-sided boundary conductance, people typically assume K=1 and multiply by 2.

**Comment by pnlfang** [comment]
Thank you! @justinmcgrath

**Comment by justinmcgrath** [comment]
I think it would be nice to figure out how to include Boost units in BioCro. Then these unit mistakes would be difficult to make.

http://www.boost.org/doc/libs/1_64_0/doc/html/boost_units.html

**Comment by pnlfang** [comment]
@justinmcgrath In Doug Orr's paper, the range of specificity factor of Rubisco seems to be off by a factor of 20 compared to "Sc/o = 2880" reported in Bernacchi et al. 2002.

**Comment by justinmcgrath** [comment]
Can you give the full citation for Bernacchi 2002? I can't find one with specificity. I suspect the difference is units though. For Kc and Ko, Doug usually uses micromole / L, whereas Carl usually uses micromole / mol. Even though specificity is dimensionless, it makes a difference since the conversion requires solubilities of CO2 and O2, which are different for the two gases. The ratio of the solubilities is about 26.

**Comment by pnlfang** [comment]
Bernacchi, C. J., A. R. Portis, H. Nakano, S. von Caemmerer, and S. P. Long (2002), Temperature response of mesophyll conductance. Implications for the determination of Rubisco enzyme kinetics and for limitations to photosynthesis in vivo, Plant Physiol, 130(4), 1992-1998, doi:10.1104/pp.008250.

I used the value of Gstar in table 1 and calculated the specificity factor from Eq. 5.

**Comment by justinmcgrath** [comment]
Then I think the difference is the units. Carl uses molar fraction and Doug uses molarity. Other groups use pressure.

**Comment by pnlfang** [comment]
@justinmcgrath Is there an example of the ray tracing approach?

**Comment by justinmcgrath** [comment]
The units in c3photoC should now all be consistent and the changes are in new-framework, but not master. I used the Boost Units library for c3photo, which I think is a nice tool for scientific programming. It would be nice to use it everywhere, and to use only SI units, instead of constantly converting prefixes.
